### PR TITLE
Implement -fsave-optimization-record

### LINKED ISSUE #2089: Enhancement: add flag to save LLVM optimization record
Implement `-fsave-optimization-record` similar as to what Clang has now. Looks like a very straightforward addition to LDC.

See here: http://llvm.org/devmtg/2016-11/Slides/Nemet-Compiler-assistedPerformanceAnalysis.pdf

Matt Godbolt's Compiler Explorer has support for it: http://lists.llvm.org/pipermail/llvm-dev/2017-May/112578.html

Implementation hints:
- Use `LLVMContext::setDiagnosticsOutputFile`
- For inspiration: look at Clang source of how setDiagnosticsOutputFile is called, and what functionality is behind `fsave-optimization-record`.

### PR DESCRIPTION
**Pull request by JohanEngelen**
Resolves #2089 .

### CODE REVIEW

File: driver/codegenerator.cpp [hunk 1]
```diff
@@ -188,7 +234,14 @@ void CodeGenerator::writeAndFreeLLModule(const char *filename) {
       {llvm::MDString::get(ir_->context(), Version)};
   IdentMetadata->addOperand(llvm::MDNode::get(ir_->context(), IdentNode));
 
+  std::unique_ptr<llvm::tool_output_file> diagnosticsOutputFile =
+      createAndSetDiagnosticsOutputFile(*ir_, context_, filename);
```
Comment (JohanEngelen) [inline driver/codegenerator.cpp]: I am not sure about this. Other suggestions for where else to place this?

**Comment by JohanEngelen** [comment]
Any concerns?

**Comment by kinke** [comment]
Looks good to me and seems like a very neat feature. As soon as it works online, be sure to announce it. ;) @rainers: VisualD supporting this would be pretty darn cool. :]

**Comment by rainers** [comment]
> VisualD supporting this would be pretty darn cool. :]

Adding the option is easy, but I guess you mean having the visual feedback as in the slides. Not so easy ;-)

**Comment by JohanEngelen** [comment]
Having the visual feedback inside Visual Studio is quite some work I imagine. But using llvm/utils/opt-viewer/opt-viewer.py generates a nice webpage that shows the optimization remarks.

**Comment by rainers** [comment]
>  But using llvm/utils/opt-viewer/opt-viewer.py generates a nice webpage that shows the optimization remarks.

I tried that but the python script failed with undecipherable errors (after installing the missing modules):
```
C:\Users\rainer\Documents\Visual Studio 14\Projects\ConsoleApp5\x64\Debug>C:\l\Python34\python.exe C:\s\d\ldc\llvm\utils\opt-viewer\opt-viewer.py opt.report html
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
For faster parsing, you may want to install libYAML for PyYAML
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\l\Python34\lib\multiprocessing\pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "C:\l\Python34\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "C:\s\d\ldc\llvm\utils\opt-viewer\optrecord.py", line 178, in get_remarks
    if not hasattr(remark, 'DebugLoc') or remark.key in all_remarks:
  File "C:\s\d\ldc\llvm\utils\opt-viewer\optrecord.py", line 120, in key
    for (key, value) in arg.iteritems():
AttributeError: 'dict' object has no attribute 'iteritems'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\s\d\ldc\llvm\utils\opt-viewer\opt-viewer.py", line 229, in <module>
    all_remarks, file_remarks, should_display_hotness = optrecord.gather_results(pmap, args.yaml_files)
  File "C:\s\d\ldc\llvm\utils\opt-viewer\optrecord.py", line 195, in gather_results
    remarks = pmap(get_remarks, filenames)
  File "C:\l\Python34\lib\multiprocessing\pool.py", line 260, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "C:\l\Python34\lib\multiprocessing\pool.py", line 599, in get
    raise self._value
AttributeError: 'dict' object has no attribute 'iteritems'
```
Any idea what's wrong here? If it worked, the build script could generate the html files and provide a clickable link that opens them within Visual Studio.

**Comment by JohanEngelen** [comment]
Looks like an LLVM bug (python2 vs python3, works for me with python2). Does the answer here fix it? `https://stackoverflow.com/questions/37803745/attributeerror-dict-object-has-no-attribute-iteritems-flask-sqlalchemy-err`

**Comment by rainers** [comment]
Thanks, using Python 2.7 works.

One problem I noticed is that the files in the report have the directory stripped from the name, so the source code annotation is not shown for code generated from imported templates or modules in sub packages (without listing all possible directories with -source-dir). Is there an LLVM option that allows keeping the full path?

**Comment by JohanEngelen** [comment]
Could also be that we don't give the full path to LLVM somewhere. The full path is passed in debuginfo, right? Just checking.

**Comment by rainers** [comment]
Yes, debug info seems to be fine.

**Comment by rainers** [comment]
The directory is a separate property in DIFile record, though, e.g. `!5 = !DIFile(filename: "srcname.d", directory: "C:\5Cs\5Cd\5Cldc\5Cldc\5Ctests\5Cdebuginfo") `. Maybe it's not used when generating the report or it dislikes back-slashes.

**Pull request merged**